src/hotspot/share/utilities/bitMap.hpp

```diff
@@ -354,84 +318,108 @@ class BitMap {
+  GrowableBitMap() : GrowableBitMap(nullptr, 0) {}
+  GrowableBitMap(bm_word_t* map, idx_t size_in_bits) : BitMap(map, size_in_bits) {}
 
  public:
-  ResourceBitMap() : BitMap(NULL, 0) {}
-  // Conditionally clears the bitmap memory.
-  ResourceBitMap(idx_t size_in_bits, bool clear = true);
-
-  // Resize the backing bitmap memory.
-  //
-  // Old bits are transferred to the new memory
-  // and the extended memory is cleared.
-  void resize(idx_t new_size_in_bits);
-
   // Set up and clear the bitmap memory.
   //
   // Precondition: The bitmap was default constructed and has
-  // not yet had memory allocated via resize or initialize.
-  void initialize(idx_t size_in_bits);
+  // not yet had memory allocated via resize or (re)initialize.
+  void initialize(idx_t size_in_bits, bool clear = true) {
+    assert(map() == NULL, "precondition");
+    assert(size() == 0,   "precondition");
+
+    resize(size_in_bits, clear);
+  }
 
   // Set up and clear the bitmap memory.
   //
   // Can be called on previously initialized bitmaps.
-  void reinitialize(idx_t size_in_bits);
+  void reinitialize(idx_t new_size_in_bits, bool clear = true) {
+    // Remove previous bits - no need to clear
+    resize(0, false /* clear */);
+
+    initialize(new_size_in_bits, clear);
+  }
+
+  // Protected functions, that are used by BitMap sub-classes that support them.
+
+  // Resize the backing bitmap memory.
+  //
+  // Old bits are transferred to the new memory
+  // and the extended memory is cleared.
+  void resize(idx_t new_size_in_bits, bool clear = true) {
+    bm_word_t* new_map = reallocate(map(), size(), new_size_in_bits, clear);
+    update(new_map, new_size_in_bits);
+  }
+};
+
+// A concrete implementation of the "abstract" BitMap class.
+//
+// The BitMapView is used when the backing storage is managed externally.
+class BitMapView : public BitMap {
+ public:
+  BitMapView() : BitMapView(nullptr, 0) {}
+  BitMapView(bm_word_t* map, idx_t size_in_bits) : BitMap(map, size_in_bits) {}
 };
 
 // A BitMap with storage in a specific Arena.
-class ArenaBitMap : public BitMap {
+class ArenaBitMap : public GrowableBitMap<ArenaBitMap> {
+  Arena* const _arena;
+
+  NONCOPYABLE(ArenaBitMap);
+
  public:
   // Clears the bitmap memory.
-  ArenaBitMap(Arena* arena, idx_t size_in_bits);
+  ArenaBitMap(Arena* arena, idx_t size_in_bits, bool clear = true);
 
- private:
-  NONCOPYABLE(ArenaBitMap);
+  bm_word_t* allocate(idx_t size_in_words) const;
+  void free(bm_word_t* map, idx_t size_in_words) const {
+    // ArenaBitMaps currently don't free memory.
+  }
+};
+
+// A BitMap with storage in the current threads resource area.
+class ResourceBitMap : public GrowableBitMap<ResourceBitMap> {
+ public:
+  ResourceBitMap() : ResourceBitMap(0) {}
+  ResourceBitMap(idx_t size_in_bits, bool clear = true);
+
+  bm_word_t* allocate(idx_t size_in_words) const;
+  void free(bm_word_t* map, idx_t size_in_words) const {
+    // ArenaBitMaps currently don't free memory.
```

Comment:
s/ArenaBitMaps/ResourceBitMap/

Or keep the old `Don't free resource allocated arrays.` comment

Reply:
I take a note and will fix this one in the next change. thanks!